File: Analise_e_projeto_de_algoritmos/triminos.cpp
#include <bits/stdc++.h>

using namespace std;

typedef vector<vector<int>> matrix;

matrix matriz;
int p = 1; // peça

void triminos(int n, int x, int y){
    int cx = x/n * n;
    int cy = y/n * n;
    n /= 2;
    cx += n;
    cy += n;
    bool bx = x & n;
    bool by = y & n;
    matriz[cx-bx][cy-!by] = matriz[cx-bx][cy-by] = matriz[cx-!bx][cy-by] = ++p;
    if(n >= 2) {
        triminos(n, x, y);
        triminos(n, cx-bx, cy-!by);
        triminos(n, cx-bx, cy-by);
        triminos(n, cx-!bx, cy-by);
    }
}

void printMatriz(int n){
    int t = log10(n*n-1/3);
    for(int y = 0; y < n; y++) {
        cout << "\b\n";
        for(int x = 0; x < n; x++) {
            cout << matriz[x][y] << "\t";
        }
        cout << "\n\n";
    }
}

int main() {
    int n, x, y;
    cin >> n >> x >> y;
    matriz = matrix(n, vector<int>(n, 0));
    matriz[x][y] = p;
    triminos(n, x, y);

    printMatriz(n);

    return 0;
}

File: Analise_e_projeto_de_algoritmos/triminos.py
import math

def triminos(n, x, y, matriz, p):
    cx = (x // n) * n
    cy = (y // n) * n
    n //= 2
    cx += n
    cy += n
    bx = x & n
    by = y & n
    matriz[cx - bx][cy - (not by)] = matriz[cx - bx][cy - by] = matriz[cx - (not bx)][cy - by] = p + 1

    if n >= 2:
        triminos(n, x, y, matriz, p + 1)
        triminos(n, cx - bx, cy - (not by), matriz, p + 1)
        triminos(n, cx - bx, cy - by, matriz, p + 1)
        triminos(n, cx - (not bx), cy - by, matriz, p + 1)

def print_matriz(n, matriz):
    t = int(math.log10(n * n - 1 / 3))
    for y in range(n):
        print()
        for x in range(n):
            print(matriz[x][y], end="\t")
        print("\n")

def main():
    # Preciso descobrir quais são os valores de entrada
    n, x, y = map(int, input().split())
    matriz = [[0] * n for _ in range(n)]
    matriz[x][y] = 1
    triminos(n, x, y, matriz, 1)
    print_matriz(n, matriz)

if __name__ == "__main__":
    main()
